File: 04-stack/00022-Generate_Parentheses/00022-generate_parentheses.cpp
// 22: Generate Parentheses
// https://leetcode.com/problems/generate-parentheses/


#include <iostream>
#include <vector>

using namespace std;


class Solution {
public:
    // SOLUTION
    vector<string> generateParentheses(int n) {
        vector<string> result;
        generate(n, 0, 0, "", result);
        return result;
    }

    void generate(int n, int open, int close, string s, vector<string>& result) {
        if (open==n && close==n) {
            result.push_back(s);
            return;
        }
        if (open < n)
            generate(n, open+1, close, s+'(', result);
        if (open > close)
            generate(n, open, close+1, s+')', result);
    }
};


int main() {
    Solution o;
    
    // INPUT
    int n = 3;

    // OUTPUT
    auto result = o.generateParentheses(n);
    cout<<"["; for (auto v : result) cout<<"\""<<v<<"\" "; cout<<"\b]"<<endl;

    return 0;
}

File: 04-stack/00022-Generate_Parentheses/00022-generate_parentheses.py
# 22: Generate Parentheses
# https://leetcode.com/problems/generate-parentheses/


class Solution:
    # SOLUTION    
    def generateParentheses(self, n: int) -> list[str]:
        result: list[str] = []
        self.generate(n, 0, 0, "", result)
        return result

    def generate(self, n: int, open: int, close: int, s: str, result) -> None:
        if open==n and close==n:
            result.append(s)
            return
        if open < n:
            self.generate(n, open+1, close, s+'(', result)
        if open > close:
            self.generate(n, open, close+1, s+')', result)



if __name__ == "__main__":
    o = Solution()

    # INPUT
    n: int = 3

    # OUTPUT
    result = o.generateParentheses(n)
    print(result)